base/abstractarraymath.jl

```diff
@@ -93,6 +93,70 @@ function _dropdims(A::AbstractArray, dims::Dims)
 end
 _dropdims(A::AbstractArray, dim::Integer) = _dropdims(A, (Int(dim),))
 
+
+"""
+    insertdims(A; dims)
+
+Inverse of [`dropdims`](@ref); return an array with new singleton dimensions
+at every dimension in `dims`.
+
+Repeated dimensions are forbidden and the largest entry in `dims` must be
+less than or equal than `ndims(A) + length(dims)`.
+
+The result shares the same underlying data as `A`, such that the
+result is mutable if and only if `A` is mutable, and setting elements of one
+alters the values of the other.
+
+See also: [`dropdims`](@ref), [`reshape`](@ref), [`vec`](@ref).
+# Examples
+```jldoctest
+julia> x = [1 2 3; 4 5 6]
+2×3 Matrix{Int64}:
+ 1  2  3
+ 4  5  6
+
+julia> insertdims(x, dims=3)
+2×3×1 Array{Int64, 3}:
+[:, :, 1] =
+ 1  2  3
+ 4  5  6
+
+julia> insertdims(x, dims=(1,2,5)) == reshape(x, 1, 1, 2, 3, 1)
+true
+
+julia> dropdims(insertdims(x, dims=(1,2,5)), dims=(1,2,5))
+2×3 Matrix{Int64}:
+ 1  2  3
+ 4  5  6
+```
+
+!!! compat "Julia 1.12"
+    Requires Julia 1.12 or later.
+"""
+insertdims(A; dims) = _insertdims(A, dims)
+function _insertdims(A::AbstractArray{T, N}, dims::Tuple{Vararg{Int64, M}}) where {T, N, M}
+    for i in eachindex(dims)
+        1 ≤ dims[i] || throw(ArgumentError("the smallest entry in dims must be ≥ 1."))
+        dims[i] ≤ N+M || throw(ArgumentError("the largest entry in dims must be not larger than the dimension of the array and the length of dims added"))
+        for j = 1:i-1
+            dims[j] == dims[i] && throw(ArgumentError("inserted dims must be unique"))
+        end
+    end
+
+    # acc is a tuple, where the first entry is the final shape
+    # the second entry off acc is a counter for the axes of A
+    inds= Base._foldoneto((acc, i) ->
+                            i ∈ dims
+                                ? ((acc[1]..., Base.OneTo(1)), acc[2])
+                                : ((acc[1]..., axes(A, acc[2])), acc[2] + 1),
+                            ((), 1), Val(N+M))
```

Comment:
```suggestion
                            ((), 1), Val((N::Int)+M))
```

Reply:
Where was this an issue?
I checked with @code_warntype, looked ok

Reply:
Try `code_warntype(_insertdims, Tuple{AbstractArray,Tuple{Int}})`.

Reply:
With these type annotations, the inferred return type of `_insertdims(::AbstractArray, ::Tuple{Int})` is still `Any`. Some things infer slightly better within the function, but I don't see how that would have a significant impact.

A downside of this is that if someone has an AbstractArray with a non-`Int` ducktyped second parameter (e.g. Int16) it would no longer work. Converting `N` to an `Int` with `N_int = Int(N)::Int` would solve that problem, but I still don't see how adding the annotations at all is helpful.

Totally open to adding these annotations if necessary, but I try to avoid adding non-idiomatic constructs for performance improvements in the absence of specific, ideally measured, examples where they improve performance.

Reply:
> I don't see how that would have a significant impact

I don't know whether these type assertions could ever lead to a measurable performance improvement, however my approach is to let the compiler into things I know but it doesn't. Compilers are very smart these days and it's difficult to predict when exactly giving it more information will help.

> A downside of this is that if someone has an AbstractArray with a non-`Int` ducktyped second parameter (e.g. Int16) it would no longer work.

AFAIK the second parameter of `AbstractArray` is supposed to be an `Int`. Although this isn't documented. I expect putting non-`Int` there would break many things and/or cause type instability all around.

>  I try to avoid adding non-idiomatic constructs for performance improvements in the absence of specific, ideally measured, examples where they improve performance.

In general, I respect avoiding premature optimization, but as far as this case is concerned I feel that there's not any harm in adding the type assertions.

This isn't a big issue in any case.

Reply:
`_dropdims` suffers from the same issues, isn't it?

So maybe that's something to discuss separately?

Reply:
If you wish.